lib/finch/request.ex

```diff
@@ -49,18 +60,28 @@ defmodule Finch.Request do
   """
   @type body() :: iodata() | {:stream, Enumerable.t()} | nil
 
+  @typedoc """
+  Optional metadata of the request. This field can be used by libraries that build on top of Finch
+  to inject information about the request that needs to be retrieved later on, for example, from
+  handlers that consume `Finch.Telemetry` events.
+  """
+  @type metadata() :: %{optional(atom()) => term()}
```

Comment:
not sure actually if we should restrict the keys of the map to atoms? it might be useful for a library to use keys like `{:the_lib_name, :some_scoped_key}`?